### PR TITLE
Update typespec for apply_changes for schemaless changesets

### PR DESCRIPTION
**Pull request by wojtekmach**

### CODE REVIEW

File: lib/ecto/changeset.ex [hunk 1]
```diff
@@ -1072,7 +1072,7 @@ defmodule Ecto.Changeset do
       apply_changes(changeset)
 
   """
-  @spec apply_changes(t) :: Ecto.Schema.t
+  @spec apply_changes(t) :: Ecto.Schema.t | data
```
Comment (wojtekmach) [inline lib/ecto/changeset.ex]: @ericmj yup:

```elixir
iex> %HexWeb.Package{} |> Ecto.Changeset.cast(%{name: "ecto"}, ~w(name)) |> Ecto.Changeset.apply_changes
%HexWeb.Package{..., name: "ecto", ...}

iex> {%{}, %{name: :string}} |> Ecto.Changeset.cast(%{name: "ecto"}, ~w(name)) |> Ecto.Changeset.apply_changes
%{name: "ecto"}
```

Comment (ericmj) [inline lib/ecto/changeset.ex]: I see. I was confused by the `data` type not fully representing the types of the `data` field.

Comment (wojtekmach) [inline lib/ecto/changeset.ex]: @josevalim sorry, I'm not well-versed in typespecs, I was essentially cargo-culting it :)

I see data is defined as: 

```elixir
# https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/changeset.ex#L239
@type data :: map()
```

so `map` vs `map()` is different wrt `nil`?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**